src/core/or/sendme.c

```diff
@@ -50,7 +53,124 @@ get_accept_min_version(void)
                                  SENDME_ACCEPT_MIN_VERSION_MIN,
                                  SENDME_ACCEPT_MIN_VERSION_MAX);
 }
-#endif
+
+/* Return true iff the given decoded SENDME version 1 cell is valid.
+ *
+ * Validation is done by comparing the digest in the cell from the previous
+ * cell we saw which tells us that the other side has in fact seen that cell.
+ * See proposal 289 for more details. */
+static bool
+cell_v1_is_valid(const sendme_cell_t *cell)
+{
+  sendme_data_v1_t *data = NULL;
+
+  tor_assert(cell);
+
+  if (sendme_data_v1_parse(&data, sendme_cell_getconstarray_data(cell),
+                           sendme_cell_getlen_data(cell)) < 0) {
+    goto invalid;
+  }
+
+  /* XXX: Match the digest in the cell to the previous cell. Needs to be
+   * implemented that is passed to this function and compared. For this, we
+   * need #26839 that is making tor remember the last digest(s). */
+
+  /* Validated SENDME v1 cell. */
+  sendme_data_v1_free(data);
+  return 1;
+ invalid:
+  sendme_data_v1_free(data);
+  return 0;
+}
+
+/* Return true iff the given cell version can be handled or if the minimum
+ * accepted version from the consensus is known to us. */
+static bool
+cell_version_is_valid(uint8_t cell_version)
+{
+  int accept_version = get_accept_min_version();
+
+  /* Can we handle this version? */
+  if (accept_version > SENDME_MAX_SUPPORTED_VERSION) {
+    log_warn(LD_PROTOCOL, "Unable to handle SENDME version %u "
+                          "(from the consensus). We only support <= %d. "
```

Comment:
Why does this say "from the consensus"?  it's coming from a cell, not the consensus.  Maybe we shouldn't warn here, but send protocol_warn instead?

Reply:
Ah!!! Gooooood catch, it is reverse. The second value is the one from the consensus. And yes protocol warn would is better!

Fixup commit: ddad00ece03e3862